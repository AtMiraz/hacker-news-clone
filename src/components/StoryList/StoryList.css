.hacker__news__app-list-container {
  margin-left: 8rem;
  margin-right: 8rem;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-evenly;
}

.hacker__news__app-list-container > .hacker__news__app-story-card {
  margin-bottom: 1.875rem;
}
