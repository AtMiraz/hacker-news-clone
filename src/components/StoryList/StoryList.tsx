import './StoryList.css'
import { useEffect, useRef, useState } from 'react'
import { Story } from '../../types/Story'
import {
  handleFetchStories,
  handleGetFavoriteStories,
  handleAddRemoveFavorite,
} from '../../utils/utils'
import { StoryListProps } from '../../types/StoryList'
import StoryCard from '../StoryCard/StoryCard'

function StoryList({ mode = 'all', newsType = 'vuejs' }: StoryListProps) {
  const [stories, setStories] = useState<Story[]>([])
  const currentPage = useRef(0)
  const isFetching = useRef(false)
  const isFirstRender = useRef(true)
  const lastElement = useRef(null)

  function goToTop() {
    window.scrollTo({ top: 0, behavior: 'smooth' })
  }

  function handleFavorite(story: Story) {
    let newStories = stories.map((newStory) => {
      if (story.createdAt === newStory.createdAt) {
        newStory.favorite = !newStory.favorite
        handleAddRemoveFavorite(newStory)
        return newStory
      }
      return newStory
    })

    if (mode === 'favorites') {
      newStories = newStories.filter((newStory) => newStory.favorite)
    }
    setStories(newStories)
  }

  async function handlePopulateStories() {
    return handleFetchStories(newsType, currentPage.current)
  }

  function handlePopulateFavorites() {
    const favorites = handleGetFavoriteStories()
    setStories(favorites)
  }

  function handleFetchData() {
    handlePopulateStories()
      .then((data) => {
        // In case the user already has some favorites
        // Try to find the favorites to mark them.
        const favorites = handleGetFavoriteStories()
        const newStories = data.map((story) => {
          const found = favorites.find(
            (favoritedStory) =>
              story.createdAt === favoritedStory.createdAt &&
              story.storyId === favoritedStory.storyId
          )
          if (found) {
            return { ...story, favorite: true }
          }
          return story
        })

        setStories((previouStories) => [...previouStories, ...newStories])
        currentPage.current++

        isFetching.current = false
        isFirstRender.current = false
      })
      .catch((e) => {
        console.warn(e)
      })
  }

  useEffect(() => {
    const observer = new IntersectionObserver(
      (entries) => {
        if (entries[0].isIntersecting && !isFetching.current) {
          handleFetchData()
        }
      },
      { threshold: 1 }
    )

    if (lastElement.current) {
      observer.observe(lastElement.current)
    }

    return () => {
      if (lastElement.current) {
        observer.unobserve(lastElement.current)
      }
    }
  }, [lastElement, newsType])

  useEffect(() => {
    if (isFirstRender.current) return
    // Cleanup whenever mode changes or news type changes
    // Set is fetching to avoid running the handleFetchData from
    currentPage.current = 0
    // the observerUseEffect
    isFetching.current = true
    setStories([])
    if (mode === 'all') {
      handleFetchData()
    } else {
      handlePopulateFavorites()
    }
  }, [mode, newsType])

  return (
    <div
      className={`hacker__news__app-list-container ${
        mode === 'favorites' && 'extra-margin-top'
      }`}
      data-testid="list-container"
    >
      {stories.length > 0
        ? stories.map((story) => (
            <StoryCard
              story={story}
              onClick={handleFavorite}
              key={`${story.createdAt}-${mode}-${story.storyId}-${story.author}-${story.title}`}
            />
          ))
        : mode === 'favorites'
        ? 'No favorite stories yet'
        : 'Loading... '}
      <div
        className="list__container-go-to-top"
        title="Go to top"
        onClick={goToTop}
      ></div>
      <div
        ref={lastElement}
        data-testid="ref-observable"
        style={{ width: 0, height: 0, overflow: 'hidden', flex: 'row' }}
      ></div>
    </div>
  )
}

export default StoryList
