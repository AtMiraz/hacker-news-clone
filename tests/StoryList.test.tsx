import React from 'react'
import StoryList from '../src/components/StoryList/StoryList'
import { render, screen } from '@testing-library/react'
import { expect, test, describe, vi, afterAll } from 'vitest'

vi.mock('../src/utils/utils', () => {
  return {
    handleFetchStories: (...args) => {
      return [
        {
          title: 'some title',
          author: 'an author',
          createdAt: '2020-05-15T20:57:02.000Z',
          favorite: false,
          url: 'some url',
          storyId: 120192921,
        },
        {
          title: 'some title',
          author: 'an author',
          createdAt: '2020-05-16T20:57:04.000Z',
          favorite: false,
          url: 'some url',
          storyId: 1235123,
        },
      ]
    },
    handleGetFavoriteStories: () => {
      return [
        {
          title: 'some title',
          author: 'an author',
          createdAt: '2020-05-16T20:57:04.000Z',
          favorite: false,
          url: 'some url',
          storyId: 1235123,
        },
      ]
    },
    handleTimeSinceDate: (string) => string,
  }
})

describe('StoryList test', () => {
  afterAll(() => {
    vi.restoreAllMocks()
  })

  test('should render the component', async () => {
    const component = render(<StoryList />)
    await screen.findByTestId('list-container')
    expect(component).toMatchSnapshot()
    expect(component).to.exist
  })

  test('should render the correct amount of children components', async () => {
    render(<StoryList />)
    await screen.findByTestId('list-container')
    const cards = await screen.findAllByTestId('story-card-component')
    expect(cards.length).toBe(2)
  })

  test('should mark a story as favorite by default when fetching if its on the localStorage', async () => {
    render(<StoryList />)
    await screen.findByTestId('list-container')

    const favorited = await screen.getByAltText('favorited')

    expect(favorited).to.exist
  })
})
