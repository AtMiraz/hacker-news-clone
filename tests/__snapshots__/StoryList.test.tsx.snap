// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StoryList test > should render the component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="hacker__news__app-list-container false"
        data-testid="list-container"
      >
        Loading... 
        <div
          class="list__container-go-to-top"
          title="Go to top"
        />
        <div
          data-testid="ref-observable"
          style="width: 0px; height: 0px; overflow: hidden;"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="hacker__news__app-list-container false"
      data-testid="list-container"
    >
      Loading... 
      <div
        class="list__container-go-to-top"
        title="Go to top"
      />
      <div
        data-testid="ref-observable"
        style="width: 0px; height: 0px; overflow: hidden;"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
